### PR TITLE
feat(dataflowgraph): validate node connection directions and required graph node config params

### PR DESCRIPTION
**Pull request by ketkimnaik**
- Validates node connection directions: rejects graphs where a Destination node is used as from or a Source node is used as to

- Validates required graph node config params: fetches the OCI artifact YAML at apply time and errors if any required: true parameters are missing from graphSettings.configuration

- Updates dataflowgraph apply help text with a Graph node example showing the configuration key-value pair format

This is how output looks like when required param for graph node setting are not present:

image

--
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to Azure IoT Operations tooling!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

Intent for Production

- [ ] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [ ] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [ ] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [ ] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT Operations CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

### CODE REVIEW

File: azext_edge/edge/_help.py [hunk 1]
```diff
@@ -808,7 +812,10 @@ def load_iotops_help():
 
           The above example defines a graph with an MQTT source flowing through a Graph processing
           node that fans out to a Kafka destination and an OpenTelemetry destination. Graph nodes
-          reference an artifact (format: `<name>:<version>`) from a registry endpoint.
+          reference an artifact (format: `<name>:<version>`) from a registry endpoint. If the
+          artifact requires configuration parameters, supply them as a list of {"key", "value"}
+          string pairs in graphSettings.configuration — omit the field entirely when no
+          configuration is needed.
           Supported nodeTypes are: Source, Destination, and Graph. Data flow graphs support only
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/_help.py]: done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 2]
```diff
@@ -328,6 +341,46 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+    ):
+        """Fetch the OCI artifact YAML layer and validate that all required parameters are provided.
+
+        The YAML layer contains moduleConfigurations[*].parameters where each entry has a
+        'required' field. Each required parameter must appear as a {"key": ..., "value": ...}
+        entry in graphSettings.configuration.
+        """
+        registry_host = registry_endpoint_obj.get("properties", {}).get("host", "")
+        image_ref = f"{registry_host}/{artifact}"
+        artifact_info = get_oci_client().fetch_first_layer(image_ref=image_ref, cmd=self.cmd)
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 3]
```diff
@@ -328,6 +341,46 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+        for module_config in yaml_data.get("moduleConfigurations", []):
+            parameters = module_config.get("parameters", {})
+            if isinstance(parameters, dict):
+                for param_name, param_info in parameters.items():
+                    if isinstance(param_info, dict) and param_info.get("required", False):
+                        required_params.add(param_name)
+
+        if required_params:
+            configuration = graph_settings.get("configuration") or []
+            provided_keys = {item.get("key") for item in configuration if isinstance(item, dict)}
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 4]
```diff
@@ -328,6 +341,46 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+        The YAML layer contains moduleConfigurations[*].parameters where each entry has a
+        'required' field. Each required parameter must appear as a {"key": ..., "value": ...}
+        entry in graphSettings.configuration.
+        """
+        registry_host = registry_endpoint_obj.get("properties", {}).get("host", "")
+        image_ref = f"{registry_host}/{artifact}"
+        artifact_info = get_oci_client().fetch_first_layer(image_ref=image_ref, cmd=self.cmd)
+
+        yaml_data = yaml.safe_load(artifact_info.content.decode("utf-8"))
+
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 5]
```diff
@@ -328,6 +350,67 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+        get_artifact_info_cached,
+    ):
+        """Fetch the OCI artifact YAML layer and validate that all required parameters are provided.
+
+        The YAML layer contains moduleConfigurations[*].parameters where each entry has a
+        'required' field. Each required parameter must appear as a {"key": ..., "value": ...}
+        entry in graphSettings.configuration.
+        """
+        registry_host = registry_endpoint_obj.get("properties", {}).get("host", "")
+        image_ref = f"{registry_host}/{artifact}"
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 6]
```diff
@@ -328,6 +350,67 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+            provided_keys = {
+                item.get("key")
+                for item in configuration
+                if isinstance(item, dict)
+                and isinstance(item.get("key"), str)
+                and item.get("key")
+                and "value" in item
+                and item.get("value") is not None
+                and (not isinstance(item.get("value"), str) or item.get("value").strip())
+            }
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:446]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 7]
```diff
@@ -328,6 +350,67 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+                f"Graph node '{node.get('name')}' artifact '{artifact}' must contain a top-level YAML object."
+            )
+
+        required_params: set = set()
+        for module_config in yaml_data.get("moduleConfigurations", []):
+            parameters = module_config.get("parameters", {})
+            if isinstance(parameters, dict):
+                for param_name, param_info in parameters.items():
+                    if isinstance(param_info, dict) and param_info.get("required", False):
+                        required_params.add(param_name)
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 8]
```diff
@@ -162,7 +165,16 @@ def get_registry_endpoint_cached(registry_endpoint_ref: str) -> dict:
...
             return registry_endpoint_cache[registry_endpoint_ref]
 
-        self._validate_graph_nodes(graph_nodes, get_registry_endpoint_cached)
+        artifact_info_cache: dict = {}
+
+        def get_artifact_info_cached(image_ref: str):
+            if image_ref not in artifact_info_cache:
+                artifact_info_cache[image_ref] = get_oci_client().fetch_first_layer(
+                    image_ref=image_ref, cmd=self.cmd
+                )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 9]
```diff
@@ -328,6 +350,67 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+                and (not isinstance(item.get("value"), str) or item.get("value").strip())
+            }
+            missing = required_params - provided_keys
+            if missing:
+                raise InvalidArgumentValueError(
+                    f"Graph node '{node.get('name')}' artifact '{artifact}' requires configuration "
+                    f"parameter(s) {sorted(missing)} but they are not provided in "
+                    "'graphSettings.configuration'. Each required parameter must be supplied as a "
+                    '{"key": "<param-name>", "value": "<value>"} entry.'
+                )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:461]: Done

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py [hunk 10]
```diff
@@ -1075,3 +1103,369 @@ def test_dataflow_graph_apply_registry_endpoint_not_found(
...
+
+
+def test_dataflow_graph_apply_with_graph_node_no_required_params(
+    mocked_cmd,
+    mocked_responses: responses,
+    mocked_get_file_config,
+    mocker,
+):
+    """Graph node apply succeeds when the artifact has no required parameters (empty config is fine)."""
+    import copy
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py]: Done

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py [hunk 11]
```diff
@@ -1075,3 +1103,369 @@ def test_dataflow_graph_apply_registry_endpoint_not_found(
...
+
+def test_dataflow_graph_apply_with_graph_node_oci_fetch_failure(
+    mocked_cmd,
+    mocked_responses: responses,
+    mocked_get_file_config,
+    mocker,
+):
+    """Apply succeeds (skips config validation) when OCI artifact fetch fails."""
+    import copy
+    from azure.cli.core.azclierror import ValidationError
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py]: Done

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py [hunk 12]
```diff
@@ -1075,3 +1103,369 @@ def test_dataflow_graph_apply_registry_endpoint_not_found(
...
+)
+def test_dataflow_graph_apply_missing_required_config(
+    mocked_cmd,
+    mocked_responses: responses,
+    mocked_get_file_config,
+    mocker,
+    configuration,
+    expected_error_text: str,
+):
+    import copy
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 13]
```diff
@@ -328,6 +361,93 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+                        if isinstance(param_info, dict) and param_info.get("required", False):
+                            required_params.add(param_name)
+
+        if required_params:
+            configuration = graph_settings.get("configuration") or []
+            provided_keys = {
+                item.get("key")
+                for item in configuration
+                if self._is_config_entry_provided(item)
+            }
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:446]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 14]
```diff
@@ -328,6 +361,93 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+        The YAML layer contains moduleConfigurations[*].parameters where each entry has a
+        'required' field. Each required parameter must appear as a {"key": ..., "value": ...}
+        entry in graphSettings.configuration.
+        """
+        registry_host = registry_endpoint_obj.get("properties", {}).get("host", "")
+        if not isinstance(registry_host, str) or not registry_host.strip():
+            raise InvalidArgumentValueError(
+                f"Graph node '{node.get('name')}' artifact '{artifact}' references a misconfigured "
+                "registry endpoint: missing required 'properties.host'."
+            )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 15]
```diff
@@ -162,7 +168,24 @@ def get_registry_endpoint_cached(registry_endpoint_ref: str) -> dict:
...
+        artifact_info_cache: dict = {}
+
+        def get_artifact_info_cached(image_ref: str):
+            if image_ref not in artifact_info_cache:
+                try:
+                    artifact_info_cache[image_ref] = get_oci_client().fetch_first_layer(
+                        image_ref=image_ref, cmd=self.cmd
+                    )
+                except Exception as ex:
+                    logger.debug(
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 16]
```diff
@@ -162,7 +168,24 @@ def get_registry_endpoint_cached(registry_endpoint_ref: str) -> dict:
...
+                        "Failed to fetch OCI artifact '%s' — skipping client-side config validation. %s",
+                        image_ref,
+                        ex,
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 17]
```diff
@@ -4,15 +4,21 @@
...
+import yaml
+
 from typing import TYPE_CHECKING, Iterable, Optional
 
+from knack.log import get_logger
 from rich.console import Console
 
+logger = get_logger(__name__)
+
 from azure.cli.core.azclierror import InvalidArgumentValueError
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:14]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 18]
```diff
@@ -162,7 +168,25 @@ def get_registry_endpoint_cached(registry_endpoint_ref: str) -> dict:
...
+                        image_ref=image_ref, cmd=self.cmd
+                    )
+                except (AzValidationError, HttpResponseError) as ex:
+                    logger.warning(
+                        "Failed to fetch OCI artifact '%s' — skipping client-side config validation. %s",
+                        image_ref,
+                        ex,
+                        exc_info=True,
+                    )
+                    artifact_info_cache[image_ref] = None
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: The docstring now documents the intentional skip-on-failure behavior

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 19]
```diff
@@ -328,6 +362,94 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+            provided_keys = {
+                item.get("key")
+                for item in configuration
+                if self._is_config_entry_provided(item)
+            }
+            missing = required_params - provided_keys
+            if missing:
+                raise InvalidArgumentValueError(
+                    f"Graph node '{node.get('name')}' image '{image_ref}' requires configuration "
+                    f"parameter(s) {sorted(missing)} but they are not provided in "
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 20]
```diff
@@ -328,6 +362,94 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
                     f"Graph node '{node.get('name')}' has an invalid 'graphSettings.artifact' value '{artifact}'. "
                     "Expected format: '<artifact-name>:<version>'."
                 )
+            registry_host = registry_endpoint_obj.get("properties", {}).get("host", "")
+            if not isinstance(registry_host, str) or not registry_host.strip():
+                raise InvalidArgumentValueError(
+                    f"Graph node '{node.get('name')}' artifact '{artifact}' references a misconfigured "
+                    "registry endpoint: missing required 'properties.host'."
+                )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: done

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py [hunk 21]
```diff
@@ -1075,3 +1102,364 @@ def test_dataflow_graph_apply_registry_endpoint_not_found(
...
+            instance_name=instance_name,
+        ),
+        json=mock_instance_record,
+        status=200,
+    )
+    _setup_graph_node_apply_mocks(mocked_responses, instance_name, resource_group_name, "myregistry")
+
+    # Simulate OCI fetch failure — apply should proceed without error
+    mock_oci = mocker.MagicMock()
+    mock_oci.fetch_first_layer.side_effect = ValidationError("registry unreachable")
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py]: Done

**Comment by ketkimnaik** [comment]
Successful IT run: https://github.com/ketkimnaik/azure-iot-ops-cli-extension/actions/runs/25892466257/job/76098440092

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: hey since you use pattern in line 366 like `if not isinstance(registry_host, str) or not registry_host.strip():` to check whitespace, you can apply it here for `registry_endpoint_ref`

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 22]
```diff
@@ -328,6 +362,104 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
                     f"Graph node '{node.get('name')}' has an invalid 'graphSettings.artifact' value '{artifact}'. "
                     "Expected format: '<artifact-name>:<version>'."
                 )
+            registry_host = (registry_endpoint_obj or {}).get("properties", {}).get("host", "")
+            if not isinstance(registry_host, str) or not registry_host.strip():
+                raise InvalidArgumentValueError(
+                    f"Graph node '{node.get('name')}' artifact '{artifact}' references a misconfigured "
+                    "registry endpoint: missing required 'properties.host'."
+                )
+            image_ref = f"{registry_host.strip()}/{artifact}"
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: `registry_host.strip().rstrip('/')` probably be safer

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 23]
```diff
@@ -328,6 +362,104 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+                node, graph_settings, image_ref, get_artifact_info_cached
+            )
+
+    @staticmethod
+    def _is_config_entry_provided(item: dict) -> bool:
+        """Return True only when a configuration entry has a valid non-empty key and value."""
+        if not isinstance(item, dict):
+            return False
+        key = item.get("key")
+        if not isinstance(key, str) or not key:
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: `not key.strip():`

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 24]
```diff
@@ -328,6 +362,104 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+        due to network issues, auth failure, or an invalid reference), a warning is logged and
+        validation is skipped — the apply proceeds and server-side validation will catch any issues.
+        """
+        artifact_info = get_artifact_info_cached(image_ref)
+        if artifact_info is None:
+            return
+
+        try:
+            yaml_data = yaml.safe_load(artifact_info.content.decode("utf-8"))
+        except yaml.YAMLError as ex:
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: decode may have UnicodeDecodeError

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 25]
```diff
@@ -328,6 +362,104 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+        required_params: set = set()
+        module_configurations = yaml_data.get("moduleConfigurations") or []
+        if isinstance(module_configurations, list):
+            for module_config in module_configurations:
+                if not isinstance(module_config, dict):
+                    continue
+                parameters = module_config.get("parameters", {})
+                if isinstance(parameters, dict):
+                    for param_name, param_info in parameters.items():
+                        if isinstance(param_info, dict) and param_info.get("required", False):
```
Comment (HangyiWang) [inline azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: `param_info.get("required") is True`

**Review by HangyiWang — DISMISSED** [review]

**Comment by HangyiWang** [comment]
Oh one thing, do we have integration tests that cover these invalid input / validation cases? We can add some.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 25 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 24 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 23 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py (hunk 22 above)
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Agree.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Better to keep consistent.

**Comment by ketkimnaik** [comment]
> Oh one thing, do we have integration tests that cover these invalid input / validation cases? We can add some.

We we cover these cases in unit tests. Integration tests are for Happy path coverage.

**Comment by HangyiWang** [comment]
> > Oh one thing, do we have integration tests that cover these invalid input / validation cases? We can add some.
> 
> We we cover these cases in unit tests. Integration tests are for Happy path coverage.

haha okay I’m not adding any extra int test workload here. We can keep these as unit tests for now or expand integration tests later, I’m still good to merge the PR.
That said, with AI support, it’s easier to include more edge cases and validations in int tests, which can be more valuable. The main overhead is just running the tests for verification; writing them is now about as easy as unit tests, and they provide stronger coverage.

**Review by HangyiWang — DISMISSED** [review]

**Comment by ketkimnaik** [comment]
> > > Oh one thing, do we have integration tests that cover these invalid input / validation cases? We can add some.
> > 
> > 
> > We we cover these cases in unit tests. Integration tests are for Happy path coverage.
> 
> haha okay I’m not adding any extra int test workload here. We can keep these as unit tests for now or expand integration tests later, I’m still good to merge the PR. That said, with AI support, it’s easier to include more edge cases and validations in int tests, which can be more valuable. The main overhead is just running the tests for verification; writing them is now about as easy as unit tests, and they provide stronger coverage.

I have added unit tests for it :)

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 26]
```diff
@@ -4,20 +4,26 @@
...
+import yaml
+
 from typing import TYPE_CHECKING, Iterable, Optional
 
+from knack.log import get_logger
 from rich.console import Console
 
 from azure.cli.core.azclierror import InvalidArgumentValueError
-from azure.core.exceptions import ResourceNotFoundError
+from azure.cli.core.azclierror import ValidationError as AzValidationError
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 27]
```diff
@@ -328,6 +362,104 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+            if missing:
+                artifact = graph_settings.get("artifact", "")
+                artifact_description = (
+                    f"artifact '{artifact}' (resolved image '{image_ref}')"
+                    if artifact and artifact != image_ref
+                    else f"image '{image_ref}'"
+                )
+                raise InvalidArgumentValueError(
+                    f"Graph node '{node.get('name')}' {artifact_description} requires "
+                    f"configuration parameter(s) {sorted(missing)} but they are not provided in "
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 28]
```diff
@@ -162,7 +168,25 @@ def get_registry_endpoint_cached(registry_endpoint_ref: str) -> dict:
...
-        self._validate_graph_nodes(graph_nodes, get_registry_endpoint_cached)
+        artifact_info_cache: dict = {}
+
+        def get_artifact_info_cached(image_ref: str):
+            if image_ref not in artifact_info_cache:
+                try:
+                    artifact_info_cache[image_ref] = get_oci_client().fetch_first_layer(
+                        image_ref=image_ref, cmd=self.cmd
+                    )
+                except (AzValidationError, HttpResponseError) as ex:
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 29]
```diff
@@ -328,6 +362,104 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+        due to network issues, auth failure, or an invalid reference), a warning is logged and
+        validation is skipped — the apply proceeds and server-side validation will catch any issues.
+        """
+        artifact_info = get_artifact_info_cached(image_ref)
+        if artifact_info is None:
+            return
+
+        try:
+            yaml_data = yaml.safe_load(artifact_info.content.decode("utf-8"))
+        except (yaml.YAMLError, UnicodeDecodeError) as ex:
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py]: Done.

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 30]
```diff
@@ -328,6 +361,105 @@ def _validate_graph_nodes(self, graph_nodes: list, get_registry_endpoint_cached)
...
+                if self._is_config_entry_provided(item)
+            }
+            missing = required_params - provided_keys
+            if missing:
+                artifact = graph_settings.get("artifact", "")
+                artifact_description = (
+                    f"artifact '{artifact}' (resolved image '{image_ref}')"
+                    if artifact and artifact != image_ref
+                    else f"image '{image_ref}'"
+                )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:454]: Not needed.

File: azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py [hunk 31]
```diff
@@ -1096,3 +1123,780 @@ def test_dataflow_graph_apply_registry_endpoint_not_found(
...
+    ids=["ValidationError", "HttpResponseError", "ConnectionError", "Exception"],
+)
+def test_dataflow_graph_apply_with_graph_node_oci_fetch_failure(
+    mocked_cmd,
+    mocked_responses: responses,
+    mocked_get_file_config,
+    mocker,
+    fetch_side_effect,
+):
+    """Apply succeeds (skips config validation) when OCI artifact fetch raises ValidationError or HttpResponseError."""
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/tests/edge/orchestration/resources/test_dataflow_graphs_unit.py]: Done

File: azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py [hunk 32]
```diff
@@ -162,7 +167,25 @@ def get_registry_endpoint_cached(registry_endpoint_ref: str) -> dict:
...
+                    artifact_info_cache[image_ref] = get_oci_client().fetch_first_layer(
+                        image_ref=image_ref, cmd=self.cmd
+                    )
+                except Exception as ex:  # best-effort: skip validation on any fetch failure
+                    logger.warning(
+                        "Failed to fetch OCI artifact '%s' — skipping client-side config validation. %s",
+                        image_ref,
+                        ex,
+                        exc_info=True,
+                    )
```
[the remark this answers is not in the document]
Comment (ketkimnaik) [reply azext_edge/edge/providers/orchestration/resources/dataflow_graphs.py:183]: Dropped: exc_info=True

**Review by HangyiWang — APPROVED** [review]
